include/aspect/lateral_averaging.h

```diff
@@ -0,0 +1,195 @@
+       * depth slices.
+       */
+      void
+      get_composition(const unsigned int composition_index,
+                      std::vector<double> &values) const;
+
+      /**
+       * Compute a lateral average of the current viscosity
+       *
+       * @param values The output vector of laterally averaged values. The
+       * function takes the pre-existing size of this vector as the number of
+       * depth slices.
+       */
+      void
+      get_viscosity(std::vector<double> &values) const;
+
+      /**
+       * Compute a lateral average of the current velocity magnitude
+       *
+       * @param values The output vector of laterally averaged values. The
+       * function takes the pre-existing size of this vector as the number of
+       * depth slices.
+       */
+      void
+      get_velocity_magnitude(std::vector<double> &values) const;
+
+      /**
+       * Compute a lateral average of the current sinking velocity
+       *
+       * @param values The output vector of laterally averaged values. The
+       * function takes the pre-existing size of this vector as the number of
+       * depth slices.
+       */
+      void
+      get_sinking_velocity(std::vector<double> &values) const;
+
+      /**
+       * Compute a lateral average of the seismic shear wave speed: Vs
+       *
+       * @param values The output vector of laterally averaged values. The
+       * function takes the pre-existing size of this vector as the number of
+       * depth slices.
+       */
+      void
+      get_Vs(std::vector<double> &values) const;
+
+      /**
+       * Compute a lateral average of the seismic pressure wave speed: Vp
+       *
+       * @param values The output vector of laterally averaged values. The
+       * function takes the pre-existing size of this vector as the number of
+       * depth slices.
+       */
+      void
+      get_Vp(std::vector<double> &values) const;
+
+      /**
+       * Compute a lateral average of the heat flux, with the sign
+       * convention of positive heat flux when it flows upwards.
+       *
+       * @param values The output vector of laterally averaged values. The
+       * function takes the pre-existing size of this vector as the number of
+       * depth slices.
+       */
+      void
+      get_vertical_heat_flux(std::vector<double> &values) const;
+
+    private:
+
+      /**
+       * Internal routine to compute the depth average of a certain quantitiy.
+       *
+       * The functor @p fctr must be an object of a user defined type that can
+       * be arbitrary but has to satisfy certain requirements. In essence,
+       * this class type needs to implement the following interface of member
+       * functions:
+       * @code
+       * template <int dim>
+       * class Functor
+       * {
+       *   public:
+       *     // operator() will have @p in and @p out filled out if @p true
+       *     bool need_material_properties() const;
```

Comment:
I had to think hard about what the comment in front of this function means. It only became clear when I looked at the names of the arguments of `operator()` that I understood. Maybe be a bit more verbose in the comment?